assets/js/modules/analytics-4/components/setup/SetupForm.test.js

```diff
@@ -0,0 +1,459 @@
+		);
+
+		expect( finishSetup ).toHaveBeenCalledTimes( 1 );
+	} );
+
+	it( 'auto-submits the form only once in the case of an error', async () => {
+		// @TODO remove after analytics 4 is decoupled from analytics module.
+		// Currently needed in order for test to work.
+		registry.dispatch( MODULES_ANALYTICS ).setSettings( {} );
+		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
+		// @TODO remove after analytics 4 is decoupled from analytics module.
+		// Currently needed in order for test to work.
+		registry
+			.dispatch( MODULES_ANALYTICS )
+			.receiveGetProperties( [], { accountID } );
+		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetExistingTag( null );
+		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetAccountSummaries(
+			accountSummaries.map( ( account ) => ( {
+				...account,
+				propertySummaries: [],
+			} ) )
+		);
+		registry
+			.dispatch( MODULES_ANALYTICS_4 )
+			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
+				propertyIDs: [ propertyID ],
+			} );
+
+		// @TODO replace to analytics 4 datastore once it is decoupled from analytics module.
+		// Currently needed in order for test to work
+		await registry.dispatch( MODULES_ANALYTICS ).selectAccount( accountID );
+
+		// Simulate an auto-submit case where the user is returning to the page
+		// after granting extra scopes necessary to submit.
+		// In this situation, the autoSubmit is set before the user goes to oAuth,
+		// store state is snapshotted, and then restored upon returning.
+		registry
+			.dispatch( CORE_FORMS )
+			.setValues( FORM_SETUP, { autoSubmit: true } );
+		provideUserAuthentication( registry, {
+			grantedScopes: [ EDIT_SCOPE ],
+		} );
+
+		registry
+			.dispatch( CORE_USER )
+			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );
+
+		const createPropertyRegexp = new RegExp(
+			'/analytics-4/data/create-property'
+		);
+
+		fetchMock.post( createPropertyRegexp, {
+			status: 403,
+			body: {
+				code: 403,
+				error: 'Insufficient permissions',
+			},
+		} );
+
+		const finishSetup = jest.fn();
+		const { getByRole, waitForRegistry } = render(
+			<SetupForm finishSetup={ finishSetup } />,
+			{
+				registry,
+			}
+		);
+		await waitForRegistry();
+
+		// Ensure the form rendered successfully.
+		expect(
+			getByRole( 'button', { name: /Configure Analytics/i } )
+		).toBeInTheDocument();
+
+		// While not strictly needed, add waits to match the successful auto-submit test case to help avoid a false positive result.
+		await waitForRegistry();
+		await act( async () => {
+			await waitForDefaultTimeouts();
+		} );
+
+		// Create property should have only been called once.
+		expect( fetchMock ).toHaveFetchedTimes( 1, createPropertyRegexp );
+		// Setup was not successful, so the finish function should not be called.
+		expect( finishSetup ).not.toHaveBeenCalled();
+		// Expect a console error due to the API error (otherwise this test will fail).
+		expect( console ).toHaveErrored();
+	} );
+} );
```

Comment:
In order to minimize references to the legacy `analytics` datastore, may I suggest changing the test cases according to the following?

```suggestion
describe( 'SetupForm', () => {
	let registry;

	beforeEach( () => {
		registry = createTestRegistry();
		provideSiteInfo( registry );
		provideUserAuthentication( registry );
		provideModules( registry, [ { slug: 'analytics-4', active: true } ] );
	} );

	it( 'renders the form with a progress bar when GTM containers are not resolved', () => {
		provideModules( registry, [
			{ slug: 'analytics-4', active: true, connected: true },
			{ slug: 'tagmanager', active: true, connected: true },
		] );
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );

		const { container, getByRole } = render( <SetupForm />, { registry } );

		expect( container ).toMatchSnapshot();

		expect( getByRole( 'progressbar' ) ).toBeInTheDocument();
	} );

	it( 'renders the form correctly', async () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetAccountSummaries( accountSummaries );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], {
				propertyID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				{
					...defaultEnhancedMeasurementSettings,
					streamEnabled: false,
				},
				{
					propertyID,
					webDataStreamID,
				}
			);

		// @TODO: This is temporarily needed here as the AccountSelect component
		// still relies on the `analytics` datastore. This should be removed
		// once the two Analytics modules are entirely decoupled.
		registry.dispatch( MODULES_ANALYTICS ).setAccountID( accountID );

		const { container, getByText, waitForRegistry } = render(
			<SetupForm />,
			{
				registry,
			}
		);
		await waitForRegistry();

		expect( container ).toMatchSnapshot();

		expect( getByText( 'Account' ) ).toBeInTheDocument();
		expect( getByText( 'Property' ) ).toBeInTheDocument();
		expect( getByText( 'Web Data Stream' ) ).toBeInTheDocument();
	} );

	it( 'submits the form upon pressing the CTA', async () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetAccountSummaries( accountSummaries );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], {
				propertyID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				defaultEnhancedMeasurementSettings,
				{
					propertyID,
					webDataStreamID,
				}
			);

		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.selectAccount( accountID );

		// @TODO: This is temporarily needed here as the AccountSelect component
		// still relies on the `analytics` datastore. This should be removed
		// once the two Analytics modules are entirely decoupled.
		registry.dispatch( MODULES_ANALYTICS ).setAccountID( accountID );

		const finishSetup = jest.fn();
		const { getByRole, waitForRegistry } = render(
			<SetupForm finishSetup={ finishSetup } />,
			{
				registry,
			}
		);
		await waitForRegistry();

		act( () => {
			// It doesn't seem possible to show the dropdown menu within the test by clicking on the dropdown in the usual way.
			// Therefore, we simulate the click on the dropdown menu item directly, despite it being hidden.
			fireEvent.click(
				getByRole( 'menuitem', {
					name: /Example Com/i,
					hidden: true,
				} )
			);
		} );

		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );

		const updateAnalyticsSettingsRegexp = new RegExp(
			'/analytics-4/data/settings'
		);

		const dismissItemEndpointRegexp = new RegExp(
			'^/google-site-kit/v1/core/user/data/dismiss-item'
		);

		muteFetch( updateAnalyticsSettingsRegexp );

		fetchMock.post( dismissItemEndpointRegexp, {
			status: 200,
			body: JSON.stringify( [
				ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY,
			] ),
		} );

		act( () => {
			fireEvent.click(
				getByRole( 'button', { name: /Configure Analytics/i } )
			);
		} );

		await waitForRegistry();

		// An additional wait is required in order for all resolvers to finish.
		await act( waitForDefaultTimeouts );

		expect( fetchMock ).toHaveFetchedTimes(
			1,
			updateAnalyticsSettingsRegexp
		);

		expect( finishSetup ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'auto-submits the form', async () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetExistingTag( null );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetAccountSummaries(
			accountSummaries.map( ( account ) => ( {
				...account,
				propertySummaries: [],
			} ) )
		);
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], {
				propertyID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				defaultEnhancedMeasurementSettings,
				{
					propertyID,
					webDataStreamID,
				}
			);

		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.selectAccount( accountID );

		// @TODO: This is temporarily needed here as the AccountSelect component
		// still relies on the `analytics` datastore. This should be removed
		// once the two Analytics modules are entirely decoupled.
		registry.dispatch( MODULES_ANALYTICS ).setAccountID( accountID );

		// Simulate an auto-submit case where the user is returning to the page
		// after granting extra scopes necessary to submit.
		// In this situation, the autoSubmit is set before the user goes to oAuth,
		// store state is snapshotted, and then restored upon returning.
		registry
			.dispatch( CORE_FORMS )
			.setValues( FORM_SETUP, { autoSubmit: true } );

		provideUserAuthentication( registry, {
			grantedScopes: [ EDIT_SCOPE ],
		} );

		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );

		const createPropertyRegexp = new RegExp(
			'/analytics-4/data/create-property'
		);

		const createWebDataStreamRegexp = new RegExp(
			'/analytics-4/data/create-webdatastream'
		);

		const updateAnalyticsSettingsRegexp = new RegExp(
			'/analytics-4/data/settings'
		);

		const accountSummariesRegexp = new RegExp(
			'/analytics-4/data/account-summaries'
		);

		fetchMock.post( createPropertyRegexp, {
			status: 200,
			body: properties[ 0 ],
		} );

		fetchMock.post( createWebDataStreamRegexp, {
			status: 200,
			body: webDataStreams[ 2 ],
		} );

		fetchMock.get( accountSummariesRegexp, {
			status: 200,
			body: accountSummaries,
		} );

		muteFetch( updateAnalyticsSettingsRegexp );

		const finishSetup = jest.fn();
		const { getByRole, waitForRegistry } = render(
			<SetupForm finishSetup={ finishSetup } />,
			{
				registry,
			}
		);
		await waitForRegistry();

		// An additional wait is required in order for all resolvers to finish.
		await act( async () => {
			await waitForDefaultTimeouts();
		} );

		// Ensure the form rendered successfully.
		expect(
			getByRole( 'button', { name: /Configure Analytics/i } )
		).toBeInTheDocument();

		await waitForRegistry();

		// An additional wait is required in order for all resolvers to finish.
		await act( async () => {
			await waitForDefaultTimeouts();
		} );

		expect( fetchMock ).toHaveFetchedTimes( 1, createPropertyRegexp );
		expect( fetchMock ).toHaveFetchedTimes( 1, createWebDataStreamRegexp );
		expect( fetchMock ).toHaveFetchedTimes(
			1,
			updateAnalyticsSettingsRegexp
		);

		expect( finishSetup ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'auto-submits the form only once in the case of an error', async () => {
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetExistingTag( null );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetAccountSummaries(
			accountSummaries.map( ( account ) => ( {
				...account,
				propertySummaries: [],
			} ) )
		);
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );

		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.selectAccount( accountID );

		// @TODO: This is temporarily needed here as the AccountSelect component
		// still relies on the `analytics` datastore. This should be removed
		// once the two Analytics modules are entirely decoupled.
		registry.dispatch( MODULES_ANALYTICS ).setAccountID( accountID );

		// Simulate an auto-submit case where the user is returning to the page
		// after granting extra scopes necessary to submit.
		// In this situation, the autoSubmit is set before the user goes to oAuth,
		// store state is snapshotted, and then restored upon returning.
		registry
			.dispatch( CORE_FORMS )
			.setValues( FORM_SETUP, { autoSubmit: true } );

		provideUserAuthentication( registry, {
			grantedScopes: [ EDIT_SCOPE ],
		} );

		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );

		const createPropertyRegexp = new RegExp(
			'/analytics-4/data/create-property'
		);

		fetchMock.post( createPropertyRegexp, {
			status: 403,
			body: {
				code: 403,
				error: 'Insufficient permissions',
			},
		} );

		const finishSetup = jest.fn();
		const { getByRole, waitForRegistry } = render(
			<SetupForm finishSetup={ finishSetup } />,
			{
				registry,
			}
		);
		await waitForRegistry();

		// Ensure the form rendered successfully.
		expect(
			getByRole( 'button', { name: /Configure Analytics/i } )
		).toBeInTheDocument();

		// While not strictly needed, add waits to match the successful auto-submit test case to help avoid a false positive result.
		await waitForRegistry();
		await act( async () => {
			await waitForDefaultTimeouts();
		} );

		// Create property should have only been called once.
		expect( fetchMock ).toHaveFetchedTimes( 1, createPropertyRegexp );
		// Setup was not successful, so the finish function should not be called.
		expect( finishSetup ).not.toHaveBeenCalled();
		// Expect a console error due to the API error (otherwise this test will fail).
		expect( console ).toHaveErrored();
	} );
} );
```

Reply:
Changing the tests to lessen the `analytics` module involvement would kinda beat the goal of cloning the `analytics-4` components and datastore, and having that 1:1 copy that should work as before after the switch is made. As these tests should be able to work as is after `analytics` is removed from codebase and this test file, along with replaced datastore in few instances. Besides, 3 out of 5 tests even when modified per suggestion above, are still failing due to dependency on `analytics` data, and are still making a request to fetch it.

Reply:
Thank you for sharing your thoughts.

> Changing the tests to lessen the `analytics` module involvement would kinda beat the goal of cloning the `analytics-4` components and datastore, and having that 1:1 copy that should work as before after the switch is made. As these tests should be able to work as is after `analytics` is removed from codebase and this test file, along with replaced datastore in few instances.

I absolutely understand. It's just that we want to reduce dependency on the legacy `analytics` module as much as possible at this point and make the implementation as close as possible to the final expectation to reduce the burden on #7932.

If you take a look at my suggestion, this is how the test suite will ultimately turn out to be when #7932 is implemented. The suggestion practically removes the fetch requests by the `analytics` `selectAccount` action and its side effects, as they'll ultimately not exist when we no longer use this module.

> Besides, 3 out of 5 tests even when modified per suggestion above, are still failing due to dependency on `analytics` data, and are still making a request to fetch it.

I just ran a test with my suggestions applied, and all of them seem to pass, even without any changes to the snapshots. See screenshot:

![image](https://github.com/google/site-kit-wp/assets/20284937/bf533e59-ac7c-46ec-805a-c828ce7d4b20)

Please let me know what you think, thanks!

CC: @jimmymadon as co-lead, to share thoughts on my suggestions. Thanks!

Reply:
Makes sense, updating. Originally when I opened the file, didn;t notice it is the version from old `analytics` module 🤦